File: android/src/main/kotlin/video/pano/rtc/native/RtcAnnotation.kt
package video.pano.rtc.native

import com.pano.rtc.api.PanoAnnotation
import com.pano.rtc.api.RtcWbView
import video.pano.rtc.native.api.IRtcAnnotation
import video.pano.rtc.native.events.RtcAnnotationEvent
import video.pano.rtc.native.utils.EnumValueConvert
import video.pano.rtc.native.utils.EnumValueConvert.StaticParams.getAnnoScalingMode
import video.pano.rtc.native.utils.EnumValueConvert.StaticParams.getWBFontStyle
import video.pano.rtc.native.utils.EnumValueConvert.StaticParams.getWBRoleType
import video.pano.rtc.native.utils.EnumValueConvert.StaticParams.getWBToolType

class RtcAnnotation(
        private val emit: (methodName: String, data: Map<String, Any?>?) -> Unit
) : IRtcAnnotation {

    private var annotation: PanoAnnotation? = null

    fun setInnerAnnotation(annotationId: String, annotation: PanoAnnotation?) {
        this.annotation = annotation
        annotation?.setCallback(RtcAnnotationEvent(annotationId) { methodName, data ->
            emit(methodName, data)
        })
    }

    override fun startAnnotation(params: Map<String, *>, callback: Callback) {
        val view = params["view"] as RtcWbView
        callback.success(annotation?.startAnnotation(view))
    }

    override fun stopAnnotation(callback: Callback) {
        callback.success(annotation?.stopAnnotation())
    }

    override fun setVisible(params: Map<String, *>, callback: Callback) {
        val visible = params["visible"] as Boolean
        callback.success(annotation?.setVisible(visible))
    }

    override fun setRoleType(params: Map<String, *>, callback: Callback) {
        val type = (params["type"] as Number).toInt()
        callback.success(annotation?.setRoleType(getWBRoleType(type)))
    }

    override fun setToolType(params: Map<String, *>, callback: Callback) {
        val type = (params["type"] as Number).toInt()
        callback.success(annotation?.setToolType(getWBToolType(type)))
    }

    override fun setLineWidth(params: Map<String, *>, callback: Callback) {
        val size = (params["size"] as Number).toInt()
        callback.success(annotation?.setLineWidth(size))
    }

    override fun setColor(params: Map<String, *>, callback: Callback) {
        val colorMap = params["color"] as Map<*, *>
        val red = (colorMap["red"] as Number).toFloat()
        val green = (colorMap["green"] as Number).toFloat()
        val blue = (colorMap["blue"] as Number).toFloat()
        val alpha = (colorMap["alpha"] as Number).toFloat()
        callback.success(annotation?.setColor(red, green, blue, alpha))
    }

    override fun setFillType(params: Map<String, *>, callback: Callback) {
        val type = (params["type"] as Number).toInt()
        callback.success(annotation?.setFillType(EnumValueConvert.getWBFillType(type)))
    }

    override fun setFillColor(params: Map<String, *>, callback: Callback) {
        val colorMap = params["color"] as Map<*, *>
        val red = (colorMap["red"] as Number).toFloat()
        val green = (colorMap["green"] as Number).toFloat()
        val blue = (colorMap["blue"] as Number).toFloat()
        val alpha = (colorMap["alpha"] as Number).toFloat()
        val color = (alpha * 255.0f + 0.5f).toInt() shl 24 or
                ((red * 255.0f + 0.5f).toInt() shl 16) or
                ((green * 255.0f + 0.5f).toInt() shl 8) or
                (blue * 255.0f + 0.5f).toInt()
        callback.success(annotation?.setFillColor(color))
    }

    override fun setFontStyle(params: Map<String, *>, callback: Callback) {
        val style = (params["style"] as Number).toInt()
        callback.success(annotation?.setFontStyle(getWBFontStyle(style)))
    }

    override fun setFontSize(params: Map<String, *>, callback: Callback) {
        val size = (params["size"] as Number).toInt()
        callback.success(annotation?.setFontSize(size))
    }

    override fun clearContents(params: Map<String, *>, callback: Callback) {
        callback.success(annotation?.clearContents())
    }

    override fun clearUserContents(params: Map<String, *>, callback: Callback) {
        val userId = (params["userId"] as String).toLong()
        callback.success(annotation?.clearUserContents(userId))
    }

    override fun undo(callback: Callback) {
        callback.success(annotation?.undo())
    }

    override fun redo(callback: Callback) {
        callback.success(annotation?.redo())
    }

    override fun snapshot(params: Map<String, *>, callback: Callback) {
        val outputDir = params["outputDir"] as String
        callback.success(annotation?.snapshot(outputDir))
    }

    override fun getToolType(callback: Callback) {
        callback.success(annotation?.toolType?.value)
    }

    override fun setAspectSize(params: Map<String, *>, callback: Callback) {
        val width = (params["width"] as Number).toInt()
        val height = (params["height"] as Number).toInt()
        callback.success(annotation?.setAspectSize(width, height))
    }

    override fun setScalingMode(params: Map<String, *>, callback: Callback) {
        val mode = (params["mode"] as Number).toInt()
        callback.success(annotation?.setScalingMode(getAnnoScalingMode(mode)))
    }

    override fun setOption(params: Map<String, *>, callback: Callback) {
        when ((params["type"] as Number).toInt()) {
            1 -> { //EnableLocalRender
                val option = params["option"] as Boolean
                callback.success(annotation?.enableLocalRender(option))
            }
            2 -> { //EnableShowDraws
                val option = params["option"] as Boolean
                callback.success(annotation?.enableShowDraws(option))
            }
            3 -> { //EnableUIResponse
                val option = params["option"] as Boolean
                callback.success(annotation?.enableUIResponse(option))
            }
            4 -> { //EnableCursorposSync
                val option = params["option"] as Boolean
                callback.success(annotation?.enableCursorPosSync(option))
            }
            5 -> { //EnableShowRemoteCursor
                val option = params["option"] as Boolean
                callback.success(annotation?.enableShowRemoteCursor(option))
            }
        }
    }
}

File: ios/Classes/Src/RtcAnnotation.swift
//
//  RtcAnnotation.swift
//  pano_rtc
//
//  Copyright © 2022 Pano. All rights reserved.
//

import Foundation
import PanoRtc

protocol RtcAnnotationInterface {
    func setRoleType(_ params: NSDictionary, _ callback: Callback)
    
    func startAnnotation(_ params: NSDictionary, _ callback: Callback)
    
    func stopAnnotation(_ params: NSDictionary, _ callback: Callback)
    
    func setVisible(_ params: NSDictionary, _ callback: Callback)
    
    func setToolType(_ params: NSDictionary, _ callback: Callback)
    
    func setLineWidth(_ params: NSDictionary, _ callback: Callback)
    
    func setColor(_ params: NSDictionary, _ callback: Callback)
    
    func setFillType(_ params: NSDictionary, _ callback: Callback)
    
    func setFillColor(_ params: NSDictionary, _ callback: Callback)
    
    func setFontStyle(_ params: NSDictionary, _ callback: Callback)
    
    func setFontSize(_ params: NSDictionary, _ callback: Callback)
    
    func undo(_ params: NSDictionary, _ callback: Callback)
    
    func redo(_ params: NSDictionary, _ callback: Callback)
    
    func clearUserContents(_ params: NSDictionary, _ callback: Callback)
    
    func clearContents(_ params: NSDictionary, _ callback: Callback)
    
    func snapshot(_ params: NSDictionary, _ callback: Callback)
    
    func getToolType(_ params: NSDictionary, _ callback: Callback)
    
    func setAspectSize(_ params: NSDictionary, _ callback: Callback)
    
    func setScalingMode(_ params: NSDictionary, _ callback: Callback)
    
    func setOption(_ params: NSDictionary, _ callback: Callback)
}

protocol RtcAnnotationDelegate: AnyObject {
    func emit(_ methodName: String, _ data: Dictionary<String, Any?>?)
}

@objc
class RtcAnnotationCache: NSObject, RtcAnnotationInterface {
    weak var delegate: RtcAnnotationDelegate?
    private var annotationMap = [String: PanoRtcAnnotation]()
    private var annotationDelegateMap = [String: PanoRtcAnnotationDelegateHandler]()
    
    func cleanup() {
        annotationMap.removeAll()
        annotationDelegateMap.removeAll()
    }
    
    subscript(annotationId: String) -> PanoRtcAnnotation? {
        get {
            return annotationMap[annotationId]
        }
    }
    
    func create(_ annotationId: String, _ annotation: PanoRtcAnnotation) {
        let annotationDelegate = PanoRtcAnnotationDelegateHandler() { [weak self] methodName, data in
            var newData: [String: Any?] = ["annotationId": annotationId]
            newData.merge(data ?? [:]) { $1 }
            self?.delegate?.emit(methodName, newData)
        }
        annotation.setDelegate(annotationDelegate)
        annotationMap[annotationId] = annotation
        annotationDelegateMap[annotationId] = annotationDelegate
    }
    
    @objc func setRoleType(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["annotationId"] as! String]?.setRoleType(PanoWBRoleType(rawValue: params["type"] as! Int)!))
    }
    
    @objc func startAnnotation(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["annotationId"] as! String]?.start(params["view"] as? UIView))
    }
    
    @objc func stopAnnotation(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["annotationId"] as! String]?.stop())
    }
    
    @objc func setVisible(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["annotationId"] as! String]?.setVisible(params["visible"] as! Bool))
    }
    
    @objc func setToolType(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["annotationId"] as! String]?.setToolType(PanoWBToolType(rawValue: params["type"] as! Int)!))
    }
    
    @objc func setLineWidth(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["annotationId"] as! String]?.setLineWidth(params["size"] as! UInt32))
    }
    
    @objc func setColor(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["annotationId"] as! String]?.setColor(PanoWBColor(map: params["color"] as! Dictionary)))
    }
    
    @objc func setFillType(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["annotationId"] as! String]?.setFillType(PanoWBFillType(rawValue: params["type"] as! Int)!))
    }
    
    @objc func setFillColor(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["annotationId"] as! String]?.setFill(PanoWBColor(map: params["color"] as! Dictionary)))
    }
    
    @objc func setFontStyle(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["annotationId"] as! String]?.setFontStyle(PanoWBFontStyle(rawValue: params["style"] as! Int)!))
    }
    
    @objc func setFontSize(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["annotationId"] as! String]?.setFontSize(params["size"] as! UInt32))
    }
    
    @objc func undo(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["annotationId"] as! String]?.undo())
    }
    
    @objc func redo(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["annotationId"] as! String]?.redo())
    }
    
    @objc func clearUserContents(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["annotationId"] as! String]?.clearUserContents(UInt64(params["userId"] as! String) ?? 0))
    }
    
    @objc func clearContents(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["annotationId"] as! String]?.clearContents())
    }
    
    @objc func snapshot(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["annotationId"] as! String]?.snapshot(params["outputDir"] as! String))
    }
    
    @objc func getToolType(_ params: NSDictionary, _ callback: Callback) {
        callback.resolve(self[params["annotationId"] as! String]) { $0.getToolType() }
    }
    
    @objc func setAspectSize(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["annotationId"] as! String]?.setAspectWidth(params["w"] as! UInt32, height: params["h"] as! UInt32))
    }
    
    @objc func setScalingMode(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["annotationId"] as! String]?.setScalingMode(PanoVideoScalingMode(rawValue: params["mode"] as! Int)!))
    }
    
    @objc func setOption(_ params: NSDictionary, _ callback: Callback) {
        var option: NSObject? = nil
        let type = PanoAnnoOptionType(rawValue: params["type"] as! Int)
        var isValid = true
        switch type {
        case .enableLocalRender:
            option = NSNumber(value: params["option"] as! Bool)
        case .showDraws:
            option = NSNumber(value: params["option"] as! Bool)
        case .enableUIResponse:
            option = NSNumber(value: params["option"] as! Bool)
        case .cursorPosSync:
            option = NSNumber(value: params["option"] as! Bool)
        case .showRemoteCursor:
            option = NSNumber(value: params["option"] as! Bool)
        default:
            isValid = false
        }
        if isValid {
            callback.code(self[params["annotationId"] as! String]?.setOption(option, for: type!))
        } else {
            callback.code(PanoResult.invalidArgs)
        }
    }
}
